src/System.Net.Http/src/System/Net/Http/Managed/HttpConnectionPool.cs

```diff
@@ -2,130 +2,440 @@
+                    // pools that have recently been used but currently aren't; if a pool was used since the last
+                    // time we cleaned up, give it another chance.
+                    if (_associatedConnectionCount == 0 && !_usedSinceLastCleanup)
+                    {
+                        Debug.Assert(list.Count == 0, $"Expected {nameof(list)}.{nameof(list.Count)} == 0");
+                        _disposed = true;
+                        return true; // Pool is disposed of.  It should be removed.
+                    }
                 }
+
+                // Reset the cleanup flag.  Any pools that are empty and not used since the last cleanup
+                // will be purged next time around.
+                _usedSinceLastCleanup = false;
             }
 
-            public bool TryAdd(HttpConnection connection)
+            // Pool is active.  Should not be removed.
+            return false;
+        }
+
+        /// <summary>A cached idle connection and metadata about it.</summary>
+        [StructLayout(LayoutKind.Auto)]
+        private struct CachedConnection : IEquatable<CachedConnection>
+        {
+            /// <summary>The cached connection.</summary>
+            internal readonly HttpConnection _connection;
+            /// <summary>The last time the connection was used.</summary>
+            internal readonly DateTimeOffset _returnedTime;
+
+            /// <summary>Initializes the cached connection and its associated metadata.</summary>
+            /// <param name="connection">The connection.</param>
+            public CachedConnection(HttpConnection connection)
             {
-                HashSet<HttpConnection> set = _sets[connection.GetHashCode() & s_slotsMask];
-                lock (set) return set.Add(connection);
+                _connection = connection;
+                _returnedTime = DateTimeOffset.UtcNow;
             }
 
-            public bool TryRemove(HttpConnection connection)
+            /// <summary>Gets whether the connection is currently usable.</summary>
+            /// <param name="now">The current time.  Passed in to ammortize the cost of calling DateTime.UtcNow.</param>
+            /// <returns>
+            /// true if we believe the connection can be reused; otherwise, false.  There is an inherent race condition here,
+            /// in that the server could terminate the connection or otherwise make it unusable immediately after we check it,
+            /// but there's not much difference between that and starting to use the connection and then having the server
+            /// terminate it, which would be considered a failure, so this race condition is largely benign and inherent to
+            /// the nature of connection pooling.
+            /// </returns>
+            public bool IsUsable(DateTimeOffset now) =>
+                now - _returnedTime <= TimeSpan.FromMilliseconds(MaxIdleTimeMilliseconds) &&
```

Comment:
For the cleanup sweep, yes, that's where stale connections are removed.  This is used both in that code path and when getting a connection from the pool.  I guess we could avoid the timeout check on that latter path. Seems reasonable.